homeassistant/components/recorder/migration.py

```diff
@@ -1879,7 +1932,39 @@ class EntityIDMigration(BaseRunTimeMigration):
 
     required_schema_version = STATES_META_SCHEMA_VERSION
     migration_id = "entity_id_migration"
-    task = EntityIDMigrationTask
+    task = CommitBeforeMigrationTask
+    # We have to commit before to make sure there are
+    # no new pending states_meta about to be added to
+    # the db since this happens live
+
+    @staticmethod
+    def migrate_data(instance: Recorder) -> bool:
+        """Migrate some data, returns True if migration is completed."""
+        return migrate_entity_ids(instance)
+
+    def migration_done(self, instance: Recorder, did_migrate: bool) -> None:
+        """Will be called after migrate returns True."""
+        # The migration has finished, now we start the post migration
+        # to remove the old entity_id data from the states table
+        # at this point we can also start using the StatesMeta table
+        # so we set active to True
+        _LOGGER.debug("Activating states_meta manager as all data is migrated")
+        instance.states_meta_manager.active = True
+        if did_migrate:
+            instance.queue_task(EntityIDPostMigrationTask())
+        else:
+            with contextlib.suppress(SQLAlchemyError):
+                # If ix_states_entity_id_last_updated_ts still exists
+                # on the states table it means the entity id migration
+                # finished by the EntityIDPostMigrationTask did not
+                # because they restarted in the middle of it. We need
+                # to pick back up where we left off.
+                if get_index_by_name(
+                    self.session,
+                    TABLE_STATES,
+                    LEGACY_STATES_ENTITY_ID_LAST_UPDATED_INDEX,
+                ):
+                    instance.queue_task(EntityIDPostMigrationTask())
```

Comment:
Not sure if we need the if/else here, or if we should just always run the `get_index_by_name` to simplify the code?
Also, maybe the `EntityIDPostMigrationTask` should be handled by another `BaseRunTimeMigration` class, which would store the removal of the index in the `MigrationChanges` table so we don't need this special case?

Reply:
@bdraco WDYT?

Reply:
> Not sure if we need the if/else here, or if we should just always run the `get_index_by_name` query to simplify the code?

I think we can simplify it to remove the branch and run the `get_index_by_name` unconditionally every time the migrator runs in case they shutdown in the middle of a migration it always checks.  The check if an index exists is cheap.
> 
> Also, maybe the `EntityIDPostMigrationTask` should be handled by another `BaseRunTimeMigration` class, which would store the removal of the index in the `MigrationChanges` table so we don't need this special case?
> 
> Or, we let this migrator class be responsible both for the data migration and the post migration task?

It would be nice to use the same migrator base class for `EntityID` and `EventId` migrations

Reply:
> It would be nice to use the same migrator base class for EntityID and EventId migrations

OK, I'll do that in a follow-up then

Reply:
> I think we can simplify it to remove the branch and run the get_index_by_name unconditionally every time the migrator runs in case they shutdown in the middle of a migration it always checks. The check if an index exists is cheap.

Done